internal/dot/graph.go

```diff
@@ -20,53 +20,166 @@
 
 package dot
 
-import "fmt"
+import (
+	"fmt"
+	"reflect"
+)
 
 // Ctor encodes a constructor provided to the container for the DOT graph.
 type Ctor struct {
-	Name    string
-	Package string
-	File    string
-	Line    int
-	Params  []*Node
-	Results []*Node
+	Name       string
+	Package    string
+	File       string
+	Line       int
+	Params     []*Param
+	GroupParam []*Group
+	Results    []*Result
+}
+
+// Param is a parameter node in the graph.
+type Param struct {
+	*Node
+	Optional bool
+}
+
+// Result is a result node in the graph.
+type Result struct {
+	*Node
+	GroupIndex int
+}
+
+// Group is a group node in the graph.
+type Group struct {
+	Type    reflect.Type
+	Group   string
+	Results []*Result
 }
 
 // Graph is the DOT-format graph in a Container.
 type Graph struct {
-	Ctors []*Ctor
+	Ctors  []*Ctor
+	Groups map[key]*Group
 }
 
-// Node is a single node in a graph.
+// Node is a single node in a graph and is embedded into Params and Results.
 type Node struct {
-	Type     string
-	Name     string
-	Optional bool
-	Group    string
+	Type  reflect.Type
+	Name  string
+	Group string
+}
+
+type key struct {
+	t reflect.Type
+	g string
+}
+
+// NewGraph creates an empty graph.
+func NewGraph() *Graph {
+	return &Graph{
+		Groups: make(map[key]*Group),
+	}
+}
+
+// NewGroup creates a new group with information in the key.
+func NewGroup(k key) *Group {
+	return &Group{
+		Type:  k.t,
+		Group: k.g,
+	}
+}
+
+// AddCtor adds the constructor with paramList and resultList into the graph.
+func (dg *Graph) AddCtor(c *Ctor, paramList []*Param, resultList []*Result) {
+	var params []*Param
+	var groupParam []*Group
+
+	// Loop through the paramList to separate them into regular params and
+	// grouped params. For grouped params, we use getGroup to find the actual
+	// group.
+	for _, param := range paramList {
+		if param.Group != "" {
+			k := key{t: param.Type.Elem(), g: param.Group}
+			group := dg.getGroup(k)
+			groupParam = append(groupParam, group)
+		} else {
+			params = append(params, param)
+		}
+	}
+
+	for _, result := range resultList {
+		// If the result is a grouped value, we want to update its GroupIndex
+		// and add it to the Group.
+		if result.Group != "" {
+			k := key{t: result.Type, g: result.Group}
+			group := dg.getGroup(k)
+
+			result.GroupIndex = len(group.Results)
+			group.Results = append(group.Results, result)
```

Comment:
nit: This is fine as is, but we should consider encapsulating all of this in another helper `dg.addGroup(k, result)`, `dg.addGroupNode(result)` or something similar. That way we aren't mutating the `group.Results` slice outside of the method call.

Reply:
I added a new `dg.addToGroup(*result)` function; however we would still have to mutate the Result passed in to update the `GroupIndex`.